packages/react-native/src/storage/Chunkifier.ts

```diff
@@ -0,0 +1,120 @@
+export type ChunkSplitterFactory<W extends Chunk> = () => ChunkSplitter<W>;
+
+/**
+ * Implementation of splitter should return either one or two `Buffer`s.
+ *
+ * The first `Buffer` will be written to the current chunk.
+ * If the second `Buffer` is returned, `chunkifier` will create a new chunk and write the
+ * second buffer to the new chunk.
+ */
+export type ChunkSplitter<W extends Chunk> = (chunk: W) => [W, W?];
+
+/**
+ * Implementation of chunk sink should return each time a new writable stream.
+ *
+ * `n` determines which stream it is in sequence.
+ */
+export type ChunkSink<W extends Chunk, S extends WritableStream<W> = WritableStream<W>> = (n: number) => S;
+
+export type Chunk = { readonly length: number };
+
+export interface ChunkifierOptions<W extends Chunk> {
+    /**
+     * Chunk splitter factory. The factory will be called when creating a new chunk.
+     */
+    readonly splitter: ChunkSplitterFactory<W>;
+
+    /**
+     * Chunk sink. The sink will be called when creating a new chunk.
+     */
+    readonly sink: ChunkSink<W>;
+
+    readonly allowEmptyChunks?: boolean;
+}
+
+interface StreamContext<W extends Chunk> {
+    readonly stream: WritableStream<W>;
+    readonly streamWriter: WritableStreamDefaultWriter<W>;
+    isEmptyChunk: boolean;
+}
+
+export class ChunkifierSink<W extends Chunk> implements UnderlyingSink<W> {
+    private _context?: StreamContext<W>;
+    private _splitter?: ChunkSplitter<W>;
+    private _chunkCount = 0;
+
+    constructor(private readonly _options: ChunkifierOptions<W>) {}
+
+    public async write(data: W): Promise<void> {
+        // If data is empty from the start, forward the write directly to current stream
+        if (this.isEmpty(data)) {
+            return await this.ensureStreamContext().streamWriter.write(data);
```

Comment:
do we need to run a write operation at all? Based on the stream writer implementation, we don't need to?

Reply:
It's just that any write to this sink should end up in writing something to the underlying stream.